kythe/go/localrun/cmd/localrun/main.go

```diff
@@ -0,0 +1,205 @@
+	indexingTimeout = flag.Duration("indexing_timeout", 300*time.Second, "How long to wait before indexing a compilation unit times out")
+	cacheSize       = datasize.Flag("cache_size", "3gb", "How much ram to dedicate to handling")
+)
+
+var errNotEnoughArgs = errors.New("not enough arguments")
+
+func getTargets(args []string) ([]string, error) {
+	if len(args) < 1 {
+		return nil, fmt.Errorf("You provided %d arguments but expected 2: %w", len(args), errNotEnoughArgs)
+	}
+	return args, nil
+}
+
+// usage prints out usage information for the program to stderr.
+func usage() {
+	fmt.Fprintf(os.Stderr, `localrun
+
+localrun is a small binary that can be used to build, extract, index, and serve any bazel repo.
+
+Usage:
+
+localrun [--language=a,b,c] [targets+]
+
+Flag arguments:
+
+`)
+	flag.Usage()
+
+	fmt.Fprintf(os.Stderr, `
+Positional arguments:
+
+  [targets+]: A list of one or more Bazel build targets that should be included in processing.
+
+Sample invocations:
+
+"bazel build" and generate cross references for all languages for everything:
+
+	localrun //...
+
+"bazel build" a portion of the repository and generate cross references for all languages:
+
+	localrun //just/this/path/...
+
+"bazel build" a portion of the repository (including non-go code) and generate cross references
+for just the go code:
+
+	localrun --language=go //just/this/path/...
+
+"bazel build" and generate cross references for just go and java code in the whole repo:
+
+	localrun --language=go,java //...
+
+Since this uses a standard bazel build invocation to build the targets, any targets that are
+included in the path will be included. If you would like to limit the build to just java_library
+targets you can employ bazel query to this end.
+
+Build only java_library targets and cross reference only java code:
+
+bazel query 'kind(java_library, //...)' | xargs localrun
+`)
+}
+
+// main entrypoint for the program.
+func main() {
+	flag.Parse()
+
+	*workingDir, _ = os.Getwd()
+	cacheDir, _ := os.UserCacheDir()
+	*outputDir = filepath.Join(cacheDir, "output")
+
+	if *publicResources == "" {
+		*publicResources = *kytheRelease + "/resources/public"
+	}
+
+	targets, err := getTargets(flag.Args())
+	if err != nil {
+		usage()
+		log.Fatalf("Error invoking localrun: %v", err)
+	}
+
+	log.Printf("Building %v for targets: %s\n",
+		languages.LanguageSet.String(), strings.Join(targets, " "))
+
+	if *workerPoolSize == 0 {
+		*workerPoolSize = runtime.GOMAXPROCS(0)
+	}
+
+	r := &localrun.Runner{
+		KytheRelease: *kytheRelease,
+		WorkingDir:   *workingDir,
+		OutputDir:    *outputDir,
+		CacheSize:    cacheSize,
+
+		WorkerPoolSize: *workerPoolSize,
+
+		Languages: languages.LanguageSet,
+		Targets:   targets,
+
+		Timeout: *indexingTimeout,
+
+		Port:            *port,
+		Hostname:        *hostname,
+		PublicResources: *publicResources,
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```

Comment:
Add a comment for why this cancellation is necessary.

Reply:
It's to prevent run-away goroutines, but I'm happy to remove it if you don't think that's a good idea.